juju/juju
# httpserver: enable using a different port for controller connections

## Description of change

For controllers with a lot of agents we see problems at controller agent start time (for example after upgrades) - as soon as the controller opens its API port it's hammered by connections from the agents and doesn't manage to get up and stable properly. To fix this, we enable setting a controller API port (`controller-api-port` in controller config) that will be used by the controller agents to connect to themselves and the other controllers, and the regular API port won't accept requests until the controllers are connected. There's also another option `api-port-open-delay` (set as a duration, eg `20s`) - if this is set the API port won't accept connections for that long after the controllers are ready.

The new options can be changed in a running controller (otherwise big existing controllers wouldn't be able to use them). To support this there's a new `controller-port` worker which bounces when the `controller-api-port` option is changed - the http-server and peer-grouper depend on this so that they'll respond to the changed value.

The firewaller will need to be changed to open the controller API port if it's set, and some providers that don't use the firewaller but still need to manage ports will also need updating. These will happen in follow-up PRs.

## QA steps

* Bootstrap a controller, enable-ha, deploy some applications.
* Once everything is up, set `controller-api-port` to 17071 and `api-open-port-delay` to 20s.
* Everything should settle again.
* Check in debug-log that the http-servers on each controller machine report that they open the controller ports first and then the normal API port after a 20s delay. 
* The api-callers in each agent should be using the right port (based on whether they're a controller agent or not). The connection delay should also be reflected in the agent logs.

## Documentation changes

Yes - the new controller config options `controller-api-port` and `api-port-open-delay` will need documentation.
